File: exercise/ts/day08/tests/doubles/inMemoryToyRepository.ts
import {Toy} from "../../src/domain/toy";
import {ToyRepository} from "../../src/domain/toyRepository";

export class InMemoryToyRepository implements ToyRepository {
    private toys: Toy[] = [];

    findByName(name: string): Toy | null {
        return this.toys.find(t => t.getName() === name) || null;
    }

    save(toy: Toy): void {
        this.toys = this.toys.filter(t => t.getName() !== toy.getName());
        this.toys.push(toy);
    }
}

File: solution/day08/src/test/java/doubles/InMemoryToyRepository.java
package doubles;

import domain.Toy;
import domain.ToyRepository;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class InMemoryToyRepository implements ToyRepository {
    private final List<Toy> toys = new ArrayList<>();

    @Override
    public Optional<Toy> findByName(String name) {
        return toys.stream()
                .filter(t -> t.getName().equals(name))
                .findFirst();
    }

    public Toy getUnsafe(String name) {
        return findByName(name)
                .orElseThrow();
    }

    @Override
    public void save(Toy toy) {
        toys.remove(toy);
        toys.add(toy);
    }
}